README.md
# Simple Cloudfront invalidator
[![Build
Status](https://secure.travis-ci.org/laurilehmijoki/simple-cloudfront-invalidator.png)]
(http://travis-ci.org/laurilehmijoki/simple-cloudfront-invalidator)

A library for invalidating AWS Cloudfront items.

## Usage

This gem is meant to be used as a dependency in other Ruby projects.

Take a look at <https://github.com/laurilehmijoki/cf-s3-invalidator>, which
makes it possible to invalidate S3-based Cloudfront distributions from the
command-line interface.

## Tests

Run `rake test`.


changelog.md
# Changelog

This project uses [Semantic Versioning](http://semver.org)

## 1.1.0

* Digest::Digest is deprecated; use Digest

## 1.0.1

* Use UTC timestamp to sign request

## 1.0.0

* First release


spec/test_helper.rb
require 'vcr'
require 'require_relative'
require_relative '../lib/simple-cloudfront-invalidator'

VCR.configure do |config|
  config.cassette_library_dir = "fixtures/cassettes"
  config.hook_into :webmock
end


spec/cloudfront_invalidator_spec.rb
require 'test_helper'

module CloudfrontInvalidator
  describe CloudfrontClient do
    it "calls the Cloudfront invalidation API" do
      VCR.use_cassette('cloudfront', :match_requests_on => [:host]) do
        @client = CloudfrontClient.new('some aws key', 'aws secret', 'cloudfrontdistid')
        report = @client.invalidate(['index.html', 'articles.html'])
        report.join('').should include("Invalidating Cloudfront items...")
        report.join('').should include("/index.html")
        report.join('').should include("/articles.html")
        report.join('').should include("succeeded")
      end
    end
  end
end


spec/simple_cloudfront_invalidator_spec.rb
require 'test_helper'

module SimpleCloudfrontInvalidator
  describe CloudfrontClient do
    before(:each) do
      VCR.use_cassette('cloudfront', :match_requests_on => [:host]) do
        @client = CloudfrontClient.new('some aws key', 'aws secret', 'cloudfrontdistid')
        @report = @client.invalidate(['index.html', 'articles.html'])
      end
    end

    it "calls the Cloudfront invalidation API" do
      expect(@report).to_not be nil
    end

    it 'returns a textual report on the invalidation' do
      textual_report = @report[:text_report].join('')
      expect(textual_report).to include "Invalidating Cloudfront items..."
      expect(textual_report).to include "/index.html"
      expect(textual_report).to include "/articles.html"
      expect(textual_report).to include "succeeded"
    end

    it 'returns the count of successfully invalidated items' do
      count = @report[:invalidated_items_count]
      expect(count).to be 2
    end
  end
end


lib/cloudfront-invalidator.rb
require 'rubygems'
require 'openssl'
require 'digest/sha1'
require 'net/https'
require 'base64'
require 'colored'

module CloudfrontInvalidator
  class CloudfrontClient
    def initialize(aws_account, aws_secret, distribution)
      @aws_account = aws_account
      @aws_secret = aws_secret
      @distribution = distribution
    end

    def invalidate(items)
      items = prefix_with_slash(items)
      body = %|
        <InvalidationBatch>
          <Paths>
            <Quantity>#{items.length}</Quantity>
            <Items>
              #{to_xml items}
            </Items>
          </Paths>
          <CallerReference>#{Time.now.utc.to_i}</CallerReference>
        </InvalidationBatch>
      |
      res = sign_and_call(
        "https://cloudfront.amazonaws.com/2012-05-05/distribution/#{@distribution}/invalidation",
        Net::HTTP::Post, body)
      create_report(items)
    end

    private

    def sign_and_call(url, method, body = nil)
      date = Time.now.strftime("%a, %d %b %Y %H:%M:%S %Z")
      digest = Base64.encode64(
        OpenSSL::HMAC.digest(OpenSSL::Digest::Digest.new('sha1'), @aws_secret, date)).strip
      uri = URI.parse(url)
      req = method.new(uri.path)
      req.initialize_http_header({
        'x-amz-date' => date,
        'Content-Type' => 'text/xml',
        'Authorization' => "AWS %s:%s" % [@aws_account, digest]
      })
      req.body = body unless body == nil
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = true
      http.verify_mode = OpenSSL::SSL::VERIFY_NONE
      res = http.request(req)
      if res.code.to_i.between? 200, 299
        res
      else
        raise "AWS API call failed. Reason:".red + "\n" + res.body
      end
    end

    def create_report(items)
      report = []
      report << "Invalidating Cloudfront items..."
      items.each do |item|
        report << "  #{item}".yellow
      end
      report << "succeeded".green
    end

    def prefix_with_slash(file_names)
      file_names.map { |file_name| "/#{file_name}" }
    end

    def to_xml(items)
      items.map { |item| "<Path>#{item}</Path>" }
    end
  end
end


lib/simple-cloudfront-invalidator.rb
require 'rubygems'
require 'openssl'
require 'digest/sha1'
require 'net/https'
require 'base64'
require 'colored'

module SimpleCloudfrontInvalidator
  class CloudfrontClient
    def initialize(aws_account, aws_secret, distribution)
      @aws_account = aws_account
      @aws_secret = aws_secret
      @distribution = distribution
    end

    def invalidate(items)
      items = prefix_with_slash(items)
      body = %|
        <InvalidationBatch>
          <Paths>
            <Quantity>#{items.length}</Quantity>
            <Items>
              #{to_xml items}
            </Items>
          </Paths>
          <CallerReference>#{Time.now.utc.to_i}</CallerReference>
        </InvalidationBatch>
      |
      res = sign_and_call(
        "https://cloudfront.amazonaws.com/2012-05-05/distribution/#{@distribution}/invalidation",
        Net::HTTP::Post, body)
      return { :text_report => create_report(items),
               :invalidated_items_count => items.length }
    end

    private

    def sign_and_call(url, method, body = nil)
      date = Time.now.utc.strftime("%a, %d %b %Y %H:%M:%S %Z")
      digest = Base64.encode64(
        OpenSSL::HMAC.digest(OpenSSL::Digest.new('sha1'), @aws_secret, date)).strip
      uri = URI.parse(url)
      req = method.new(uri.path)
      req.initialize_http_header({
        'x-amz-date' => date,
        'Content-Type' => 'text/xml',
        'Authorization' => "AWS %s:%s" % [@aws_account, digest]
      })
      req.body = body unless body == nil
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = true
      http.verify_mode = OpenSSL::SSL::VERIFY_NONE
      res = http.request(req)
      if res.code.to_i.between? 200, 299
        res
      else
        raise "AWS API call failed. Reason:".red + "\n" + res.body
      end
    end

    def create_report(items)
      report = []
      report << "Invalidating Cloudfront items..."
      items.each do |item|
        report << "  #{item}".yellow
      end
      report << "succeeded".green
    end

    def prefix_with_slash(file_names)
      file_names.map { |file_name| "/#{file_name}" }
    end

    def to_xml(items)
      items.map { |item| "<Path>#{item}</Path>" }
    end
  end
end
